### ISSUE TITLE
Please add 'pre' and 'post' hooks

### ISSUE THREAD

**Issue by lemzwerg** [Josef-Friedrich/nodetree#16]
Right now, if I select, say, the `hyph` hook, 'nodetree' completely replaces any existing hooks with its own.  This makes it inconvenient to actually watch what already installed hooks (or LuaTeX's default actions) actually do.

I thus suggest a new feature to improve this: If I add a prefix `pre-` or `post-` to a hook name, nodetree's hook gets installed as the very first or the very last, respectively.  For example, adding both `pre-hyph` and `post-hyph` would allow me closely watch what LuaTeX's default hyphenation handler is doing.

**Comment by Josef-Friedrich** [comment]
If the `luatexbase` module is available, `nodetree` uses `luatexbase` to register its callbacks. Using `luatexbase` it is possible to register multiple callbacks.

In plain LuaTex you can load `luatexbase` using `\input ltluatex.tex`

https://github.com/Josef-Friedrich/nodetree/blob/d613470202cfc32b8a6cd9bc747b2ffa48af8e48/nodetree.lua#L1551-L1560

**Comment by lemzwerg** [comment]
Yes, I could do this on the lua side.  However, it would be nice to make 'nodetree' do something like the following pseudo-code if I call `\NodetreeRegisterCallback`.

```
<get original hook>
if <pre>
  new_hook = function, which calls nodetree hook then original hook
else if <post>
  new_hook = function, which calls original hook then nodetree hook
else
  new_hook = nodetree hook
```

**Comment by lemzwerg** [comment]
By the way, the newest version of `luatexbase.add_to_callback` (as distributed with the core LaTeX stuff) allows to install stuff 'before' and 'after' hooks...

**Comment by Josef-Friedrich** [comment]
> ... the newest version of luatexbase.add_to_callback allows to install stuff 'before' and 'after' hooks...

Are you sure? As I understand it, you can only change the order of the registered user functions.

http://mirrors.ctan.org/macros/latex/base/ltluatex.pdf

`luatexbase.declare_callback_rule(⟨name⟩, ⟨first⟩, ⟨relation⟩, ⟨second ⟩)`: relation = before | after

**Comment by Josef-Friedrich** [comment]
> ```
> <get original hook>
> if <pre>
>   new_hook = function, which calls nodetree hook then original hook
> else if <post>
>   new_hook = function, which calls original hook then nodetree hook
> else
>   new_hook = nodetree hook
> ```

thanks for your thoughts in pseudo code form. I don't know how to get the original hook? Is the original hyphenate algorithm not written in C? If your registered hyphenate hook returns false, then the internal hyphenate algorithm is disabled.

From the LuaTeX manual:

> No return values. This callback has to insert discretionary nodes in the node list it receives.
> Setting this callback to false will prevent the internal discretionary insertion pass.

**Comment by lemzwerg** [comment]
I'm a LuaTeX newbie, so I'm not sure at all and just poke in the dark.  Perhaps I got this wrong.  It probably makes sense to discuss this on the LuaTeX mailing list...

**Comment by Josef-Friedrich** [comment]
I have drawn a diagram that visualizes in what order the callbacks are called:
[Callbacks graph](https://viewer.diagrams.net/?highlight=0000ff&edit=_blank&nav=1&title=luatex-callbacks.drawio#R7V1bc6u2Fv41eXSGO%2Fgxzb70dNqePSedaXtePBhkQ4MRBdlx%2BusrGTB4STYEA8I72Q%2FZsJC4rPVp3aQl3%2BmPm%2F3X1E2CX7CPojtN8fd3%2Bqc7TdMsQ6f%2FMcprQbENI6es09DPaWpFeAr%2FQQVRKajb0EfZSUOCcUTC5JTo4ThGHjmhuWmKX06brXB0%2BtTEXSOO8OS5EU%2F9PfRJkFMdU6noP6JwHZRPVpXiysYtGxeELHB9%2FFIj6Z%2Fv9McUY5IfbfaPKGLcK%2FmS9%2Fty5urxxVIUkzYd%2FL9%2B3z%2Fj5OcvP%2F76U%2Fj%2Fv7%2F%2B7GTbWSGMnRttiw%2B%2B06yI3u8HP9zRwzU7XG7DyGdfuliFEUFp2YQ%2Bq9ZK0HG3oYBY0Fcsry1T2Bre48Ap8lqyP8Xb2EfsC1R6%2BSUICXpKXI9dfaGIo7SAbKLiMn276BFHOD301X0XOSuP0jOS4mdUu2J5Dlqujs%2Br87Bg6w6lBO1rpIKnXxHeIJK%2B0ibF1ZluFAIuIK7bxflLhRe1BEFQw4pV0NwCouvjvSsp0oNCkGKhYrIzZ%2Fbm4ds%2FP4WPz%2Bbn33bffpipnFCD1yRAsUvQdQzugV3q1JjFj4AoXLtkm4bxWjq3pscus43CCOOMfuIiwlSP1od%2Fo8KI0ctinbqrcfTFAAKyVNkCsjgBPaM0ngKYDWVqYLbbgDmgPHl%2Bq%2BULlnh%2FsyA2HNmCUZ1B%2FZLgPfoltiVdqvNBpTqu8ZAj1GM0MRl7o4k8AsBYFPsPLBajZ17kZlnonfKyYrzCOPiMiBeUJyfs9NHK3UbkEjMzvE091OwdEzddI9LsGCL%2FJEDkRVPjvClgfElLUeSScHcaVoqkUTzhGw7pl1V%2BswN0tA0Emn930aseBoIb6VDZq%2BBGOWO4Gx3AcfzsK%2FDCOyiTwculELkRL%2Bqk8GLoQPsrHfFiAOBZ5sh4ETlpE8HLJb3RiBdrWngBakHrql8MC9xIGxkvItdxynixWuLFnhZeTKBfoJjb4sUEUaEFgTc0XkRO6ZTxYt%2BmPQJqwda74gVmXJxx8aLz2dVp46WtfjGnhZfvxX%2FRRVMsEvEC5HsKDGuOn%2FfEUx%2FtP7IvX%2F%2FjfQv8%2F830twDtwx5di5dbi6fLNNN7NUhAv%2Bgj%2B7v6hOPpi1nA5gSMMinAOD3hRdUlOzC8wxuFMVqmyK2mOQCAxp6RMIE2lz9VVIKxxrVNFGZkQfAiYAfSeSZzEkfoN%2FCOspskVBUxlu0OvOsPbRybBMxsjTZTOus0gU7vb6KECsGNyOu050mukig3%2BaUZskWqcyJNUrQYQvNexTk4SS9fjciMn7hgqcHhuST5Rn8nN8w37w%2BPmKARsltm%2FPRWvFx00BoDc7nxkw3wYnTFC5zSno%2BLF94dTjD1TSanmaGDZ5uyNXP5QsCm4S1JtpNx7mYqUGzqXDrfRO7dVFXURck3qyhHpo6agZh93lVHzUDM7oyso1Tee5wuYtquojgf%2BUjDiwa833nXLPJMBTGlM1waWYyYm3eb26YJ5aoYHbrNcDlV63kq6DZD8zM0YG7eby4XDdcRc14XSdMxR%2F%2Bj1DFdAy06xoGOGTnSUmVOPfSiY7SWVikfmx865lrAtCq46JzXpFh7TzlNVXpOs12Zxq5LAc1utAIaIMyV4yFPKMylY7JVBsNEpbp0WcpcB9ePMm%2B7sl9q3mzWm%2FmXbP1LdF4DmFYQ8bbp7jiYe8SLyF%2BcYBJDhyubYO6htfUHwS2XsT0DGCpB97XWLGENsgsIB8AskycV%2FvI79otG0bLMfkxRlkQhWeCVnMKycewRrDOXbY1KXTKQp%2BiuaOsiUf2eXEb5cr35tFRrq5F%2F2c1bDaAZ1GNicyw3o039VxSFSYaAbxG4Cbue0RckT4RtfsINQ%2BXwTzQMV6viSh%2FD0DDN02HomNw41AVYgK5hf8OwTZEUHS1PxSlOSYDXOHajzxW1zmzGn9Bzo4coXMeUtsSE4E0ek5eDGScozinFNk%2FORcZfGOZoH5I%2Fiiez4z8Z%2Fd4szj7ta80%2BvRYnZ1RAc1Rhn9UVXRONV2cQQew4dhpHb%2BP5XxqSDAKTGpCG5PEotaion%2BBba2kWdbmp1N7MIrSL2nB19GI%2B3r4nJahknGL2HUIG5lk6x99QnQyNmNufrmmrZHJofSDmWsTwvvd%2FU5%2FF2QpeMTkEiP6lbl%2B0dL3njN0uPpJTtEJUxB473rjx1o3OR9oVRb2nR%2B6GRc%2FxMksObdjmniQNl1sCA%2F2qnybqdy4%2FUHXTz3ZbHPrlO9nx%2FQxRP%2FGCbtjVFHVt7maJup0rqICdbeHripc5wr6OqK9oU7Sqy1zURZR7q8lduT%2Fe9m%2BKlVBUn6AydBQ3Ot%2BIQYHKAnsoyxbplqo2vo1%2BX8gLLFbkGhqH16q2z%2BQaMHEq9S0juRZMcspxEz7uMpONcm7HxFo7JgeloeaIusSkORfVgwcNq6rLJFJNwWpzkQcNvaT%2BygT4oEQw%2FGkgsX%2FIFU956LKQdcO%2Bu0Y7yQ8qtSvVJmN1am2jW6W2uVy9SalJlGP1Tf3qEq0wRaMfZknkgmv5ywgvHWa8AQbcIgSP0IphYUV1J2%2BUGbUM7nnjHeMYxmYFqQfozGCBgHAe2xCARx2uxkTO%2BkqxbPp0sVsnK6XOiWpgTlRVOu9uBtdIKMOVY4uRxC%2BjOzV8ypoCJfGojrjTH8Qaqdg89Dj%2B%2Fb%2B2GUH%2BAtBvbuTD9Woz0XTEuAP%2F5kMis22xkNTiMgsueui6gs1s2hZm6OIyfpWMIDgBCHrbRCEbn3l%2BXNXOj%2BJBBqgN403pJUaWnMRoOyF0G8ZXbeplSjXTjca17TCGBn8%2BcnWE1TV5ephMvHm0td6C5QNu%2FcCta23FO4Ob3IV23xHe5Li1EwbWB656wRW%2FbARkbL8Xz1f%2BtieWnPX4o4%2Fh1rtdWJMyDgYUfNcxPOD%2B%2B0I2lirjO8LVRV017V1ULPizDqZ9bxm2YjqOppvaXO8GMhukSQxbu2e3M1TVZre3x4Wc1hFyze5vj9g7BXWPSGydopuUl3K7Cq6HhVE9wqqH6qUrfV%2Bpmd8Z3E%2B582KYGdym0Wm5GubN9Utw746y3nrQAia7h%2F0XbgG2ArM8vdST2lQq13WZqDm2MuwaRdxc5uk60ys3uADr%2Bc2uplc22pweYosegUThkr6yOhHl3jbL8z%2BL5xxOqlKRw1lZK9JR%2BU184xn424FG582t4DZZRsvS8zdbYu5B86ssMT2tfm89b179bL3%2B%2BV8%3D)

**Comment by lemzwerg** [comment]
This looks great, but I don't understand it at all.  Please add a legend and some more documentation.

**Comment by Josef-Friedrich** [comment]
I tried to visualize in which order the callbacks are executed. The `ligaturing` hook is always called after the `hyphenate` hook.  But before the `hyphenate` hook come many different hooks.

**Comment by lemzwerg** [comment]
OK.  However, some steps need more explanation IMHO.

* It deserves a footnote or something to give an example where the 'buildpage_filter' is really called as the very first thing after input has started.
* The same for filters that point to themselves.
* Finally, there are lines with two arrows.  It's not clear what this exactly means – this could be a ping-pong interaction, or it could be meant as an abbreviation of two single-headed arrow lines.
* It would be very nice if there was a small input example, probably on a separate page, for which you demonstrate the processing route.

**Comment by lemzwerg** [comment]
The wanted functionality is now available by using leading and trailing colons to the callback name (commit ac67fe792bb37b51c21f7629e7f4f2fa4933b91b and some follow-up fixes until a52c7560a6efdd3e6abbb79867c3e2ba4670d8a9).
